src/error/error.rs

```diff
@@ -2,47 +2,133 @@ use self::KindPriv::*;
-    UnprocessableEntity,
-    Internal,
+// ===== impl Builder =====
+
+impl Builder {
+    fn new () -> Self {
+        Self {
+            kind: None,
+            detail: None,
+            status: None,
+        }
+    }
+
+    /// Set status of the error.
+    pub fn status(self, status: StatusCode) -> Self {
+        Self {
+            kind: self.kind,
+            detail: self.detail,
+            status: Some(status),
+
+        }
+    }
+
+    /// Set kind and title of the error.
+    pub fn kind(self, kind: &str, title: &str) -> Self {
+        Self {
+            kind: Some((kind.to_owned(), title.to_owned())),
+            detail: self.detail,
+            status: self.status,
+        }
+    }
+
+    /// Set detailed information about the error.
+    pub fn detail(self, detail: &str) -> Self {
+        Self {
+            kind: self.kind,
+            detail: Some(detail.to_owned()),
+            status: self.status,
+        }
+    }
+
+    /// Create an error object.
+    pub fn build(self) -> Error {
+        let mut err =
+            match (self.kind, self.status) {
+                (Some((ref kind, ref title)), Some(status)) => Error::new(kind, title, status),
+                (Some(_), None) => Error::from(StatusCode::INTERNAL_SERVER_ERROR),
+                (None, Some(status)) => Error::from(status),
+                (None, None) => Error::from(StatusCode::INTERNAL_SERVER_ERROR),
+            };
+
+        match self.detail {
+            Some(ref detail) => {
+                err.set_detail(detail);
+                err
+            },
+            None => err,
+        }
+    }
 }
 
 // ===== impl Error =====
 
 impl Error {
+    // TODO: this function should return '&self.kind' property
+    // and its return value should be changed to '&str'
     /// Returns the corresponding `ErrorKind` for this error.
+    #[deprecated(note="return value of the function will be changed to &str")]
     pub fn kind(&self) -> &ErrorKind {
-        &self.kind
+        &self.error_kind
     }
 
-    /// Returns a status code for this error.
-    pub fn status_code(&self) -> StatusCode {
-        match self.kind.kind {
-            BadRequest => StatusCode::BAD_REQUEST,
-            Unauthorized => StatusCode::UNAUTHORIZED,
-            Fordidden => StatusCode::FORBIDDEN,
-            NotFound => StatusCode::NOT_FOUND,
-            UnprocessableEntity => StatusCode::UNPROCESSABLE_ENTITY,
-            Internal => StatusCode::INTERNAL_SERVER_ERROR,
+    /// Create an error object.
+    pub fn new(kind: &str, title: &str, status: StatusCode) -> Self {
+        Self {
+            kind: kind.to_owned(),
+            title: title.to_owned(),
+            detail: None,
+            status,
+
+            // TODO: this property isn't used and should be removed
+            error_kind: ErrorKind::new(&status),
         }
     }
+
+    /// Set detailed information about the error.
+    pub fn set_detail(&mut self, value: &str) -> &mut Self {
```

Comment:
Is this still needed if there is  a builder?

Reply:
I think we can keep it, so user of the library will decide which way of creating of an error object he or she prefer.

Reply:
It is also used from within the `Builder`